pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,573 @@
+        }
+        if (!removeGroup) {
+          newReplicaGroupPartitionAssignment.setInstancesToReplicaGroup(partitionId, groupId, newReplicaGroup);
+        }
+      }
+
+      // Adding new replica groups if needed
+      int index = 0;
+      for (int newGroupId = oldNumReplicaGroup; newGroupId < targetNumReplicaGroup; newGroupId++) {
+        List<String> newReplicaGroup = new ArrayList<>();
+        while (newReplicaGroup.size() < targetNumInstancesPerPartition) {
+          newReplicaGroup.add(addedServers.get(index));
+          index++;
+        }
+        newReplicaGroupPartitionAssignment.setInstancesToReplicaGroup(partitionId, newGroupId, newReplicaGroup);
+      }
+    }
+
+    return newReplicaGroupPartitionAssignment;
+  }
+
+  /**
+   * Given the list of added/removed servers and the old and new replica group configurations, compute the
+   * type of update (e.g. replace server, add servers to each replica group, add replica groups..)
+   *
+   * @return the update type
+   */
+  private ReplicaGroupRebalanceType computeRebalanceType(int oldNumInstancesPerPartition, int oldNumReplicaGroup,
+      int targetNumReplicaGroup, int targetNumInstancesPerPartition, List<String> addedServers,
+      List<String> removedServers) {
+    boolean sameNumInstancesPerPartition = oldNumInstancesPerPartition == targetNumInstancesPerPartition;
+    boolean sameNumReplicaGroup = oldNumReplicaGroup == targetNumReplicaGroup;
+    boolean isAddedServersSizeZero = addedServers.size() == 0;
+    boolean isRemovedServersSizeZero = removedServers.size() == 0;
+
+    if (sameNumInstancesPerPartition && sameNumReplicaGroup && addedServers.size() == removedServers.size()) {
+      return ReplicaGroupRebalanceType.REPLACE;
+    } else if (sameNumInstancesPerPartition) {
+      if (oldNumReplicaGroup < targetNumReplicaGroup && !isAddedServersSizeZero && isRemovedServersSizeZero
+          && addedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.ADD_REPLICA_GROUP;
+      } else if (oldNumReplicaGroup > targetNumReplicaGroup && isAddedServersSizeZero && !isRemovedServersSizeZero
+          && removedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_REPLICA_GROUP;
+      }
+    } else if (sameNumReplicaGroup) {
+      if (oldNumInstancesPerPartition < targetNumInstancesPerPartition && isRemovedServersSizeZero
+          && !isAddedServersSizeZero && addedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.ADD_SERVER;
+      } else if (oldNumInstancesPerPartition > targetNumInstancesPerPartition && !isRemovedServersSizeZero
+          && isAddedServersSizeZero && removedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_SERVER;
+      }
+    }
+    return ReplicaGroupRebalanceType.UNSUPPORTED;
+  }
+
+  /**
+   * Rebalances segments for a replica group table without update
```

Comment:
"Modifies in-momory ideal state to Rebalance segments for table with replica-group based segment assignment."

Reply:
changed the comment